### PR TITLE
PrometheusEmitter NullPointerException fix

### PR DESCRIPTION
**Pull request by 599166320**
Fixes [#13285](https://github.com/apache/druid/issues/13285).

### Description

PrometheusEmitter NullPointerException fix

<hr>

##### Key changed/added classes in this PR
 * `PrometheusEmitter`

<hr>

This PR has:

- [X] been self-reviewed.
   - [X] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [X] been tested in a test Druid cluster.

### CODE REVIEW

File: extensions-contrib/prometheus-emitter/src/main/java/org/apache/druid/emitter/prometheus/PrometheusEmitter.java
Comment (cryptoe) [inline extensions-contrib/prometheus-emitter/src/main/java/org/apache/druid/emitter/prometheus/PrometheusEmitter.java:61]: Curious to understand the race here!

Comment (599166320) [reply extensions-contrib/prometheus-emitter/src/main/java/org/apache/druid/emitter/prometheus/PrometheusEmitter.java:61]: The existence of multiple threads calling emitMetric will modify the `identifier`. Even if the `identifier` is modified countless times, there will always be only one result, which is an `idempotent design`.
Currently only the thread of `PrometheusPushGatewayEmitter` will read the value of `identifier`. In order to guarantee the visibility of the reading thread, I use volatile to decorate the `identifier`.

In addition, there is an inefficient way to use LinkedBlockingQueue to achieve safe publishing. This is how `InfluxdbEmitter` is implemented.

**Comment by FrankChen021** [comment]
Thanks for the fix. 

From the assignment statement, I don't understand how could `identifier` be null? 

```java
identifier = (userDims.get("task") == null ? metricEvent.getHost() : (String) userDims.get("task"));
```

**Comment by 599166320** [comment]
> 

I realized that there are two conditions that may lead to `NPE` in the access identifier

1. The `identifier` may be modified by the `emitMetric` thread. Since identifier is a common object attribute, it may not be visible to the `PrometheusPushGatewayEmitter` thread

2. `emitMetric` is not called in advance, but the identifier is accessed.

**Review by cryptoe — APPROVED** [review]
LGTM. Will merge if @FrankChen021 is okay with it.

**Comment by FrankChen021** [comment]
> > 
> 
> I realized that there are two conditions that may lead to `NPE` in the access identifier
> 
> 1. The `identifier` may be modified by the `emitMetric` thread. Since identifier is a common object attribute, it may not be visible to the `PrometheusPushGatewayEmitter` thread
> 2. `emitMetric` is not called in advance, but the identifier is accessed.

Yeah, the 2nd case is the reason. 
The check logic in `pushMetric` and `flush` is a little bit chaos. Could you improve it by putting all checks together like:

```java
  private void pushMetric()
  {
    if ( pushGateway == null || identifier == null) {
       return ;
   }
  ...
```

Another point(not relevant to this PR) is, checking if namespace is null is meaningless since it's guaranteed it won't be null.

```java
  @JsonCreator
  public PrometheusEmitterConfig(
      ...
  )
  {
    ...
    this.namespace = namespace != null ? namespace : "druid";
```

**Comment by 599166320** [comment]
@FrankChen021  I debugged. If `druid.emitter.prometheus.namespace=xxx` is not configured, the value of `namespace` will be at least `druid`. In order to make the code more concise and clear, I have deleted meaningless judgments

**Comment by FrankChen021** [comment]
Thanks for the cleanup. 
We also need to remove the `Nullable` annotation from the `namespace` since it won't be null at all.

**Review by FrankChen021 — APPROVED** [review]
LGTM. Let's wait for the CI to complete.

**Pull request merged**
